src/basicfuns.jl

```diff
@@ -441,3 +441,72 @@ $(SIGNATURES)
 Compute the complementary double exponential, `1 - exp(-exp(x))`.
 """
 cexpexp(x) = -_expm1(-exp(x))
+
+#=
+this uses the identity:
+
+log(logistic(x)) = -log(1 + exp(-x))
+=#
+"""
+$(SIGNATURES)
+
+The natural logarithm of the `logistic` function, computed more
+carefully and with fewer calls than than the composition
+`log(logistic(x))`.
+
+Its inverse is the [`logitexp`](@ref) function.
+"""
+loglogistic(x::AbstractFloat) = -log1pexp(-x) #
+loglogistic(x::T) where {T<:Real} = -log1pexp(-convert(promote_type(Float64, T), x))
```

Comment:
I don't think this is needed - and it would cause undesired promotions of eg `Float32`.
```suggestion
```

Reply:
Separate dispatch for `AbstractFloat` and all other `T<:Real` was a specific design decision, the logic for which is as follows:

Negation of `x` is necessary, but is only guaranteed to have the correct behavior for floating point types. In particular, unsigned integers will always produce nonsense results for any value other than `0`. The same applies to the typemin for all signed integers, though, that affects only 1 value, rather than 2^n - 1 values in the unsigned case. Rational numbers would replicate this problem, but `Base` throws an error when one attempts the pathological negations.

Notably, the undesired promotions only occur if the `AbstractFloat` dispatch is removed; as designed, it does no promotion of floating point types.

An alternate design is:
```julia
loglogistic(x::Real) = -log1pexp(-x)
loglogistic(x::T) where {T<:Integer} = loglogistic(convert(promote_type(Float64, T), x))
loglogistic(x::T) where {U<:Integer, T<:Rational{U}} = loglogistic(convert(promote_type(Float64, T), x))
```

I can see the motivation for this (`loglogistic(x::Real) = -log1pexp(-x)`) signature, so it looks like the alternate design should be used. Perhaps the explicit `convert` and `promote_type`s could be replaced with `float`, which, while potentially ambiguous, might be closer to what users expect.

Reply:
I guess generally most functions (in this package and other stats packages) assume that the additive inverse exists, so I assume using unsigned integers will cause problems all over the place. If you are worried about it here, I think you should just use `loglogistic(x::Real) = -log1pexp(-float(x))`. Promoting with an explicit `Float64` won't be correct in general (eg as mentioned above `Float16` or `Float32` must not be converted to `Float64`).

Reply:
None of the definitions above promote `Float16` or `Float32` to `Float64`, only integers and rationals would be converted. See the example below

```julia
f(x::AbstractFloat) = x
f(x::T) where {T<:Real} = convert(promote_type(Float64, T), x)

@assert typeof(f(Float16(1.0))) === Float16
```

Moreover, `loglogistic(x::Real) = -log1pexp(-float(x))` is not necessarily correct for non-`Base` types which are `<:Real` but are not floating point types (e.g. [unums](https://en.wikipedia.org/wiki/Unum_(number_format)).

I'd rather err toward the least intrusive set of assumptions, while maintaining correctness, even if it leads to 2 more lines of code.

> most functions (in this package and other stats packages) assume that the additive inverse exists, so I assume using unsigned integers will cause problems all over the place.

Why not make the robust choice? 1 line of code is a very small burden to pay for elimination of a whole class of potential puzzles. Strictly speaking, a function should not accept unsigned integers if it just produces nonsense results.

Reply:
Ok, so I thought I found a way to satisfy both the needs:
1) the AD abstractions, which really want a single function signature, which should be `f(x::Real)`
2) correctness wrt integer and rational arguments, without warts or quirks

The answer is straightforward: hide the necessary dispatch as a helper function, so that the public interface can satisfy the requirements AD backends expect.
```julia
loglogistic(x::Real) = _loglogistic(x)
_loglogistic(x::Real) = -log1pexp(-x)
_loglogistic(x::Union{Integer, Rational}) = _loglogistic(float(x))
```

Surely we can pay 2 extra lines of code to guarantee correctness wrt all integer and rational arguments, no?